applications: nrf_desktop: Document SUIT support (#15851)
nrfconnect/sdk-nrf: applications/nrf_desktop/bootloader_dfu.rst

This option enables services on application core that communicates with secure domain on which the SUIT is executed.
You can also use :kconfig:option:`CONFIG_SUIT_DFU_CANDIDATE_PROCESSING_MINIMAL` - This configuration allows update using SUIT when no interpretation of the SUIT envelope is needed by the local domain apart from basic format checking.

SUIT also has the following options in the sysbuild configuration:

   * :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE` - Required to create the SUIT envelope.
     Turned on by default on ``nrf54h20dk`` board.
     The envelope is a container for the manifest and code or data payloads, such as CPU cores update images.
   * :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE_SEQUENCE_NUM` - The nRF Desktop application uses the sequence number to define an application version.

.. note::
   The :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE_SIGN` is disabled, so the generated SUIT envelope is not signed.

Make sure that the dedicated DFU partition called ``dfu_partition`` is defined in the DTS.
The partition is used to store the incoming SUIT envelope with update candidate.

> 

Reviewer: Remaining part of the readme seems not updated. E.g. `Background Device Firmware Upgrade`, `Update image generation`, `Image transfer over SMP`, `Update image verification and application image update` do not include part related to SUIT

.. _nrf_desktop_suit:

Software Update for Internet of Things (SUIT)
=============================================

Software Update for Internet of Things (SUIT) is a metadata format and procedure for performing Device Firmware Update (DFU).
It allows you to update and boot multiple CPU cores of the supported System on Chip (SoC).
The SUIT DFU is currently available only for the nRF54H20 SoC.
The application stores an update image in the dedicated DFU partition and triggers firmware update using the SUIT API.
The firmware update affects the firmware of both the Application and Radio cores.
The update candidates are moved directly to the corresponding image slots.

The SUIT can be used for the :ref:`background DFU <nrf_desktop_bootloader_background_dfu>` through:

   * The :ref:`nrf_desktop_config_channel` and :ref:`nrf_desktop_dfu`.
   * The :ref:`nrf_desktop_dfu_mcumgr_suit`.

For an introduction to SUIT and more detailed information, see the :ref:`ug_nrf54h20_suit_dfu`.

Configuring SUIT
----------------

To enable SUIT on the supported target, select the :kconfig:option:`CONFIG_SUIT` Kconfig option in the application configuration.

> This option enables services on application core that communicates with secure domain on which the SUIT is executed.

Author: Leaving second part as in your proposition we use twice "that"

You can also use :kconfig:option:`CONFIG_SUIT_DFU_CANDIDATE_PROCESSING_MINIMAL` - This configuration allows update using SUIT when no interpretation of the SUIT envelope is needed by the local domain apart from basic format checking.

SUIT also has the following options in the sysbuild configuration:

   * :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE` - Required to create the SUIT envelope.
     Turned on by default on ``nrf54h20dk`` board.
     The envelope is a container for the manifest and code or data payloads, such as CPU cores update images.
   * :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE_SEQUENCE_NUM` - The nRF Desktop application uses the sequence number to define an application version.

.. note::
   The :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE_SIGN` is disabled, so the generated SUIT envelope is not signed.

> Make sure that the dedicated DFU partition called ``dfu_partition`` is defined in the DTS.

Reviewer: btw. Note that nRF54H uses DTS for memory configuration seems missing

The partition is used to store the incoming SUIT envelope with update candidate.

> For details about defining a memory layout for your target, see the :ref:`nrf_desktop_memory_layout` section.

Reviewer 2: Please verify if the `nrf_desktop_memory_layout` section contains all necessary nRF54H specific information.

Author: Currently we don't have any changes in memory layout and we use default one. Removing this reference.

Reviewer 2: > At least this one seems missing. Please add it back
> 
> ```
> .. note::
>    The Partition Manager is disabled for the nRF54H Series.
>    The memory layout for this SoC Series is always defined in the DTS.
> ```

+1, please add it

  Both mentioned firmware upgrade methods are not used simultaneously by any of the configurations.
  For example, the ``nrf52840dk/nrf52840`` board in ``prj_mcuboot_smp.conf`` uses only the background DFU and does not enable the serial recovery feature.

.. _nrf_desktop_suit:

Software Update for Internet of Things (SUIT)
=============================================

Software Update for Internet of Things (SUIT) is a metadata format and procedure for performing Device Firmware Update (DFU).
It allows you to update and boot multiple CPU cores of the supported System on Chip (SoC).
The SUIT DFU is currently available only for the nRF54H20 SoC.
The application stores an update image in the dedicated DFU partition and triggers firmware update using the SUIT API.
The firmware update affects the firmware of both the Application and Radio cores.
The update candidates are moved directly to the corresponding image slots.

The SUIT can be used for the :ref:`background DFU <nrf_desktop_bootloader_background_dfu>` through:

   * The :ref:`nrf_desktop_config_channel` and :ref:`nrf_desktop_dfu`.

>    * The :ref:`nrf_desktop_dfu_mcumgr_suit`.

Reviewer: `nrf_desktop_dfu_mcumgr`

For an introduction to SUIT and more detailed information, see the :ref:`ug_nrf54h20_suit_dfu`.

Configuring SUIT
----------------

To enable SUIT on the supported target, enable the :kconfig:option:`CONFIG_SUIT` Kconfig option in the application configuration.

> This option enables services on application core that communicates with the Secure Domain on which the SUIT is executed.

Reviewer: `service` ?

You can also enable :kconfig:option:`CONFIG_SUIT_DFU_CANDIDATE_PROCESSING_MINIMAL`, because application core only stores the SUIT envelope to a dedicated partition and then it passes control to the Secure Domain.

SUIT also has the following options in the sysbuild configuration:

   * :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE` - Required to create the SUIT envelope.
     Turned on by default on ``nrf54h20dk`` board.
     The envelope is used directly as DFU update file by DFU tools.
   * :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE_SEQUENCE_NUM` - The nRF Desktop application uses the sequence number to define an application version.

.. note::
   The :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE_SIGN` is disabled, so the generated SUIT envelope is not signed.

> On ``nrf54h20dk`` board the dedicated DFU partition called ``dfu_partition`` is defined by default in the DTS.

Reviewer 3: ```suggestion
On the ``nrf54h20dk`` board, the dedicated DFU partition called ``dfu_partition`` is defined by default in the DTS.
```

> The partition is used to store the incoming SUIT envelope with update candidate.

Reviewer 3: ```suggestion
The partition is used to store the incoming SUIT envelope with an update candidate.
```

> The application relies on default memory layout defined by the board.

Reviewer 3: ```suggestion
The application relies on the default memory layout defined by the board.
```

> For details about defining a memory layout for your target, see the :ref:`nrf_desktop_memory_layout_dts` section.

Reviewer: The linked information may actually be a bit misleading for nRF54H (memory configuration for this multi-core SoC is more complicated). Do we have a guide for nRF54H memory configuration in the NCS? If so I would link there instead. Otherwise we need to either write a short guide in scope of nRF Desktop doc or inform that we rely on default memory map and not to link DTS configuration section here.

.. note::
   The Partition Manager is disabled for the nRF54H Series.

>    The memory layout for this SoC Series is always defined in the DTS.

Reviewer 3: ```suggestion
   The memory layout for these devices is always defined in the DTS.
```

>   * :ref:`nrf_desktop_dfu_mcumgr` (supported only with MCUboot bootloader)

Reviewer: Cherry-pick is not correct here. SUIT relies on `dfu_mcumgr` module already

.. note::
   Make sure to properly configure the bootloader to ensure that the build system generates the :file:`zephyr/dfu_application.zip` archive containing all of the required update images.

SUIT
----

> The :file:`DFU/root.suit` SUIT envelope is used in the SUIT DFU procedure.

Reviewer 2: maybe it would be nice to specify it is located in the build directory?
```
The :file:`<build_dir>/DFU/root.suit` SUIT envelope is used in the SUIT DFU procedure.
```

> The envelope contains both manifest and update candidates for the Application and Radio cores.

Reviewer 3: ```suggestion
The envelope contains both the manifest and the update candidates for the application and radio cores.
```

  Therefore, this module should be used for configurations that enable multiple DFU transports (for example, if a configuration also enables :ref:`nrf_desktop_dfu`).

The `nRF Connect Device Manager`_ application transfers the image update files over the Simple Management Protocol (SMP).

To perform DFU using the `nRF Connect Device Manager`_ mobile app, complete the following steps:

.. tabs::

   .. tab:: MCUboot

      .. include:: /device_guides/nrf52/fota_update.rst
         :start-after: fota_upgrades_over_ble_nrfcdm_common_dfu_steps_start
         :end-before: fota_upgrades_over_ble_nrfcdm_common_dfu_steps_end

      .. include:: /device_guides/nrf52/fota_update.rst
         :start-after: fota_upgrades_over_ble_mcuboot_direct_xip_nrfcdm_note_start
         :end-before: fota_upgrades_over_ble_mcuboot_direct_xip_nrfcdm_note_end

      .. note::
         When the :kconfig:option:`CONFIG_MCUMGR_GRP_IMG_REJECT_DIRECT_XIP_MISMATCHED_SLOT` Kconfig option is enabled in the application configuration, the device rejects the update image upload for the invalid slot.
         It is recommended to enable the option if the application uses MCUboot in the direct-xip mode.

>    .. tab:: SUIT

Reviewer 4: Would it be possible to include these steps from the nRF54H "Working with" guide (similarly to the included content for the nRF52 Series and MCUboot)? It would be best that these steps are maintained by the Charon team as it would pretty much guarantee that they are always up to date.

Author: It would be a great idea, but I didn't see any fragment which describes this process in "Device Firmware Update using SUIT" in device guide for nrf54h20.

If you know where I can find it please link it here.

Reviewer 4: Well, this is something that we should coordinate/arrange. They certainly have steps for testing the app in their sample

Reviewer 4: With beautiful pictures:

https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/suit/smp_transfer/README.html#testing

Reviewer 2: We already discussed that with Charon some time ago, here is a related ticket: https://nordicsemi.atlassian.net/browse/NCSDK-25026

      1. Generate the SUIT envelope by building your application with the FOTA support over Bluetooth Low Energy.
         You can find the generated :file:`root.suit` envelope in the :file:`<build_dir>/DFU` directory.
      #. Download the :file:`root.suit` envelope to your device.

         .. note::
            `nRF Connect for Desktop`_ does not currently support the FOTA process.

      #. Use the `nRF Connect Device Manager`_ mobile app to update your device with the new firmware.

         a. Ensure that you can access the :file:`root.suit` envelope from your phone or tablet.
         #. In the mobile app, scan and select the device to update.
         #. Switch to the :guilabel:`Image` tab and tap on :guilabel:`ADVANCED` in the upper right corner of the app.
         #. In the **Firmware Upload** section, tap the :guilabel:`SELECT FILE` button and select the :file:`root.suit` envelope.
         #. Tap the :guilabel:`START` button.
         #. Wait for the DFU to finish and then verify that the application works properly.

      .. note::

>          Support for SUIT updates is available since the following versions of the `nRF Connect Device Manager`_ mobile app:

Reviewer 3: ```suggestion
         Support for SUIT updates is available starting from the following versions of the `nRF Connect Device Manager`_ mobile app:
```

         * Version ``1.10`` on Android.

>          * Version ``1.6`` on iOS.

Reviewer 2: smp_transfer sample lists 1.9 for android and 1.6 for iOS.
Why we have different version here?

Author: I am pretty sure they haven't updated their doc but will check with Charon and post their answer here.

Reviewer 2: As far as I know there was some kind of SUIT support already done for the 1.5.3 for iOS (https://github.com/NordicSemiconductor/IOS-nRF-Connect-Device-Manager/releases/tag/1.5.3) and 1.9 for Android (https://github.com/NordicSemiconductor/Android-nRF-Connect-Device-Manager/releases/tag/1.9.0), thus why I ask.

Author: Ok talked with Charon and using 1.9 doesn't work. Problem is that 1.10 is not yet released. but we will need it anyway to make it working - so I would leave it as it is.
